pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -0,0 +1,445 @@
+     * @param batchedWriteMaxRecords Caches “write requests” for a certain number, if reach this threshold, will trig
+     *                               Bookie writes.
+     * @param batchedWriteMaxSize Caches “write requests” for a certain size of request data, if reach this threshold,
+     *                           will trig Bookie writes.
+     * @param writeMaxDelayInMillis Maximum delay for writing to bookie for the earliest request in the batch.
+     * @param batchEnabled Enable or disabled the batch feature, will use Managed Ledger directly and without batching
+     *                    when disabled.
+     */
+    public TxnLogBufferedWriter(ManagedLedger managedLedger, OrderedExecutor orderedExecutor,
+                                ScheduledExecutorService scheduledExecutorService, DataSerializer<T> dataSerializer,
+                                int batchedWriteMaxRecords, int batchedWriteMaxSize, int writeMaxDelayInMillis,
+                                boolean batchEnabled){
+        this.batchEnabled = batchEnabled;
+        this.managedLedger = managedLedger;
+        this.orderedExecutor = orderedExecutor;
+        this.scheduledExecutorService = scheduledExecutorService;
+        this.dataSerializer = dataSerializer;
+        this.batchedWriteMaxRecords = batchedWriteMaxRecords;
+        this.batchedWriteMaxSize = batchedWriteMaxSize;
+        this.writeMaxDelayInMillis = writeMaxDelayInMillis;
+        // scheduler task.
+        if (batchEnabled) {
+            this.scheduledFuture = this.scheduledExecutorService.scheduleAtFixedRate(this::trigFlush,
+                    writeMaxDelayInMillis, writeMaxDelayInMillis, TimeUnit.MICROSECONDS);
+        }
+    }
+
+    /**
+     * Append a new entry to the end of a managed ledger. All writes will be performed in the same thread.
+     * @param data data entry to be persisted.
+     * @param callback Will call {@link AsyncCallbacks.AddEntryCallback#addComplete(Position, ByteBuf, Object)} when
+     *                 add complete.
+     *                 Note: When {@link #batchEnabled} == false, the param-byteBuf of {@param callback} will be null.
+     *                 Will call {@link AsyncCallbacks.AddEntryCallback#addFailed(ManagedLedgerException, Object)} when
+     *                 add failure.
+     * @throws ManagedLedgerException
+     */
+    public void asyncAddData(T data, AsyncCallbacks.AddEntryCallback callback, Object ctx){
+        if (!batchEnabled){
+            ByteBuf byteBuf = dataSerializer.serialize(data);
+            managedLedger.asyncAddEntry(byteBuf, DisabledBatchCallback.INSTANCE, Triple.of(byteBuf, callback, ctx));
+            return;
+        }
+        if (dataSerializer.getSerializedSize(data) >= batchedWriteMaxSize){
+            doTrigFlush(true);
+            ByteBuf byteBuf = dataSerializer.serialize(data);
+            managedLedger.asyncAddEntry(byteBuf, DisabledBatchCallback.INSTANCE, Triple.of(byteBuf, callback, ctx));
```

Comment:
move this logical to`internalAsyncAddData ` is better, otherwise, it will `dataSerializer.getSerializedSize(data)` twice. if the data is bigger than the max entry size, we should return exception

Reply:
Good suggestion